### ISSUE TITLE
[API Proposal]: string/ReadOnlySpan<char> Truncate

### ISSUE THREAD

**Issue by hrrrrustic** [dotnet/runtime#85603]
### Background and motivation

For `int` overload there is a #31655 with good use-cases and motivation
<details>
<summary>Copy-paste from #31655</summary

I propose a new method on String to perform a fairly common task: Truncating a string to a maximum length. Scenarios:

1. Inserting to a limited-length database field when data loss is preferable over failing. For example, when logging a user agent string of an HTTP request.
2. Displaying a potentially long string to the user.
3. Making sure that untrusted input is not unexpectedly long when failure is not desirable.

In my experience, this is quite a common need. I often need to define a helper function for this.
Semantics:

1. A string that is less than the maximum length is unaltered.
2. The API never returns a string larger than the maximum length.
</details>

Two more from myself:
1. Scripting/Debugging - to reduce console output when you actually don't need a full string, just a beginning to make sure it looks fine
2. This is just a shortcut for `string.Substring(0, maxLength)` or `span.Slice(0, maxLength)` without manually passing `0` 😄

Key difference from #31655 that there is no `truncationSuffix`. It is too specific for core library and probably should be placed in UI frameworks (also discussed in original issue)

I also added ROS<char> version to avoid allocating. Also keep in mind that ROS even doesn't have `Enumerable.Take` workaround and should be manually `if`-checked for length

`char` overload is basically 
```csharp
var index = str.IndexOf('');
if (index != -1)
    str = str.Substring(0, index);
```

### API Proposal

```csharp
namespace System
{
    public class String
    {
        public string Truncate(int maxLength);
        public string Truncate(char until);
    }

    public readonly ref struct Span<T>
    {
        public Span<T> Truncate(int maxLength);
    }

    public static class MemoryExtensions
    {
        public Span<char> Truncate(this Span<char> span, char until);
    }
}
```

### API Usage

Main use-cases described above, I just link some examples from /runtime. Most of them are for ROS<> since BCL tries to avoid using `.Substring`, but obviously it's not possible to use ROS<> everywhere instead string
For `int` overload `Math.Min` pattern is pretty common:
(More cases in other Asn files)
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Formats.Asn1/src/System/Formats/Asn1/AsnDecoder.cs#L548-L549
(More cases in this and other Tar files)
https://github.com/dotnet/runtime/blob/d9e86731ca1262eb246530d9d8950f94f03d8214/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs#L807-L808

https://github.com/dotnet/runtime/blob/d260e5023b6f1593fb38534c81ff030805e9ae0f/src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs#L3351-L3359

**Comment by DaZombieKiller** [comment]
The span version of `Truncate` would be more useful if it worked for any equatable `T`:
```diff
public static class MemoryExtensions
{
-   public ReadOnlySpan<char> Truncate(this ReadOnlySpan<char> span, char until);
+   public ReadOnlySpan<T> Truncate<T>(this ReadOnlySpan<T> span, T until)
+       where T : IEquatable<T>;
}
```
I often have a similar method in projects, it's useful when dealing with buffers containing null-terminated strings.

**Comment by MichalPetryka** [comment]
Maybe add an overload that takes maxLength and a value, so that it truncates to the last occurance of the `value` when the length is above the limit? I've used such helper in my code to truncate text for display in UI.

**Comment by hrrrrustic** [comment]
Updated a proposal with generic overload
I hope someone will see this issue 😄

**Comment by tannergooding** [comment]
I don't think this proposal is in the right shape. It differs quite a bit from other APIs we have on the span types.

In general, we want most methods to be extension methods in `MemoryExtensions` not instance methods on `Span/ROSpan`.

I'm not sure we want to take this proposal in as a general purpose one. Truncate can have different meanings depending on context. Consider for example what a user might expect it to mean for `ReadOnlySpan<float>`.

There is also a concern around taking both `char` and `int` since this can be very tricky due to implicit conversions.

**Comment by DaZombieKiller** [comment]
> There is also a concern around taking both `char` and `int` since this can be very tricky due to implicit conversions.

It's also ambiguous when `T` is `int` for the `Span<T>` versions.

**Comment by hrrrrustic** [comment]
> It's also ambiguous when T is int for the Span<T> versions.

Thanks! Totally forgot about that.
Well, let's just drop 'char' overload to make this issue easier.

```C#
public static class MemoryExtensions
{
    // Same on Span<T>?
    public static ReadOnlySpan<T> Truncate<T>(this ReadOnlySpan<T> span, int maxLength);
}

public class String
{
    public string Truncate(int maxLength);
}
```

>Truncate can have different meanings depending on context

What about simple `Take`? Just same as `Enumerable.Take` (and Kotlin version), but more efficient

**Comment by tannergooding** [comment]
Going to close this one. There's going to be an effort to mirror more of Enumerable APIs on Span where it makes sense, but it will be handled as a group.
